bindings/python/manifold3d.cpp

```diff
@@ -459,111 +486,90 @@ PYBIND11_MODULE(manifold3d, m) {
-          [](Mesh &self) {
-            const int numVert = self.vertNormal.size();
-            py::array_t<float> vert_normal({numVert, 3});
-            auto vert_normal_view = vert_normal.mutable_unchecked<2>();
+            if (mergeToVert.has_value()) out.mergeToVert = mergeToVert.value();
+
+            if (runIndex.has_value()) out.runIndex = runIndex.value();
+
+            if (runOriginalID.has_value())
+              out.runOriginalID = runOriginalID.value();
 
-            for (int i = 0; i < numVert; ++i)
-              for (const int j : {0, 1, 2})
-                vert_normal_view(i, j) = self.vertNormal[i][j];
-            return vert_normal;
-          })
-      .def_property_readonly("halfedge_tangent", [](Mesh &self) {
-        const int numEdge = self.halfedgeTangent.size();
-        py::array_t<float> halfedge_tangent({numEdge, 4});
-        auto halfedge_tangent_view = halfedge_tangent.mutable_unchecked<2>();
+            if (runTransform.has_value()) {
+              auto runTransform1 = runTransform.value();
+              if (runTransform1.ndim() != 3 || runTransform1.shape(1) != 4 ||
+                  runTransform1.shape(2) != 3)
+                throw std::runtime_error(
+                    "Invalid run_transform shape, expected (-1, 4, 3)");
+              runTransform1 = runTransform1.reshape({-1});
+              out.runTransform = std::vector<float>(
+                  runTransform1.data(),
+                  runTransform1.data() + runTransform1.size());
+            }
+
+            if (faceID.has_value()) out.faceID = faceID.value();
 
-        for (int i = 0; i < numEdge; ++i)
-          for (const int j : {0, 1, 2, 3})
-            halfedge_tangent_view(i, j) = self.halfedgeTangent[i][j];
-        return halfedge_tangent;
-      });
+            if (halfedgeTangent.has_value()) {
+              auto halfedgeTangent1 = halfedgeTangent.value();
+              if (halfedgeTangent1.ndim() != 3 ||
+                  halfedgeTangent1.shape(1) != 3 ||
+                  halfedgeTangent1.shape(2) != 4)
+                throw std::runtime_error(
+                    "Invalid halfedge_tangent shape, expected (-1, 3, 4)");
+              halfedgeTangent1 = halfedgeTangent1.reshape({-1});
+              out.halfedgeTangent = std::vector<float>(
+                  halfedgeTangent1.data(),
+                  halfedgeTangent1.data() + halfedgeTangent1.size());
+            }
+
+            return out;
+          }),
+          py::arg("vert_properties"), py::arg("tri_verts"),
+          py::arg("merge_from_vert") = py::none(),
+          py::arg("merge_to_vert") = py::none(),
+          py::arg("run_index") = py::none(),
+          py::arg("run_original_id") = py::none(),
+          py::arg("run_transform") = py::none(),
+          py::arg("face_id") = py::none(),
+          py::arg("halfedge_tangent") = py::none(), py::arg("precision") = 0)
+      .def_readonly("vert_properties", &MeshGL::vertProperties)
```

Comment:
So, I think this is using the basic STL interop by doing a copy, right? Should we instead do like on the input side and construct a `py::array` from the pointer and apply the correct shape?

Reply:
yes